### PR TITLE
Misc fixes and refactors

### PR DESCRIPTION
**Pull request by pauloue**
- fix missing buttons in Sfinv tab
- fix a crash by cleaning unregistered items from the saved inv_items
- apply filters in /craft
- remove type="shapeless" from single-item recipes (it can be confusing in the guide if they are displayed as shapeless since they cannot be shaped)
- move some functions to where they are registered

### CODE REVIEW

File: init.lua [hunk 1]
```diff
@@ -587,16 +590,6 @@ local function item_in_inv(item, inv_items)
 	end
 end
 
-local function init_data(player, name)
```
Comment (kilbith) [inline init.lua]: Keep this function.

File: init.lua [hunk 2]
```diff
@@ -607,22 +600,6 @@ local function reset_data(data)
 	data.items       = data.items_raw
 end
 
-local function get_init_items()
```
Comment (kilbith) [inline init.lua]: Keep this function.

File: init.lua [hunk 3]
```diff
@@ -56,13 +56,13 @@ local function table_merge(t, t2)
...
+local function clean_items(items)
 	local hash, ct = {}, {}
-	for i = 1, #t do
-		local v = t[i]
-		if not hash[v] then
-			ct[#ct + 1] = v
-			hash[v] = true
+	for i = 1, #items do
+		local item = items[i]
+		if not hash[item] and reg_items[item] then
```
Comment (kilbith) [inline init.lua]: > `reg_items[item]`

You are sure this isn't checked before?

**Comment by kilbith** [comment]
> * move some functions to where they are registered

No, refused.

File: init.lua [hunk 4]
```diff
@@ -910,57 +908,61 @@ if progressive_mode then
...
+		reset_data(data)
+
+		local recipes = recipes_cache[node_name]
+		local is_fuel = fuel_cache[node_name]
+
+		if recipes then
+			recipes = apply_recipe_filters(recipes, player)
+		end
+
+		if not recipes or #recipes == 0 then
```
Comment (kilbith) [inline init.lua]: Btw we need some standardization to check an empty array. Either we pick `next(table)` or `#recipes == 0` but no both mixed in the codebase.

Comment (pauloue) [reply init.lua]: `#recipes` is faster, but it can only be used for array tables.

**Pull request merged**